agent/aws_iid_attestor.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/hcl"
+
+	"github.com/spiffe/spire/proto/agent/nodeattestor"
+	"github.com/spiffe/spire/proto/common"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+
+	aia "github.com/spiffe/aws-iid-attestor/common"
+)
+
+const (
+	pluginName                  = "aws_iid_attestor"
+	defaultIdentityDocumentUrl  = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+	defaultIdentitySignatureUrl = "http://169.254.169.254/latest/dynamic/instance-identity/signature"
+)
+
+type IIDAttestorConfig struct {
+	TrustDomain          string `hcl:"trust_domain"`
+	IdentityDocumentUrl  string `hcl:"identity_document_url"`
+	IdentitySignatureUrl string `hcl:"identity_signature_url"`
+}
+
+type IIDAttestorPlugin struct {
+	trustDomain          string
+	identityDocumentUrl  string
+	identitySignatureUrl string
+
+	awsAccountId  string
+	awsInstanceId string
+
+	mtx *sync.RWMutex
+}
+
+func (p *IIDAttestorPlugin) spiffeID() *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, p.awsAccountId, p.awsInstanceId)
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host:   p.trustDomain,
+		Path:   spiffePath,
+	}
+	return id
+}
+
+func httpGetBytes(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (p *IIDAttestorPlugin) FetchAttestationData(req *nodeattestor.FetchAttestationDataRequest) (*nodeattestor.FetchAttestationDataResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	docBytes, err := httpGetBytes(p.identityDocumentUrl)
+	if err != nil {
+		err = aia.AttestationStepError("retrieving the IID from AWS", err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+
+	var doc aia.InstanceIdentityDocument
+	err = json.Unmarshal(docBytes, &doc)
+	if err != nil {
+		err = aia.AttestationStepError("unmarshaling the IID", err)
+		return &nodeattestor.FetchAttestationDataResponse{}, err
+	}
+
+	p.awsAccountId = doc.AccountId
+	p.awsInstanceId = doc.InstanceId
```

Comment:
We should avoid storing these on the struct, since they are specific to a particular attestation request, and not related to long-term function of the plugin. We would additionally need write lock on mutex here rather than read lock.

Looks like this is only used to form the SPIFFE ID, so perhaps that method can take these as args instead